### PR TITLE
enhance: [cherry-pick] Alter compactTo segments before compactFrom to avoid data loss if crash

### PR DESCRIPTION
**Pull request by wayblink**
issue: #34512 
pr: #34513

### CODE REVIEW

File: internal/datacoord/meta.go [hunk 1]
```diff
@@ -1367,92 +1391,81 @@ func (m *meta) completeClusterCompactionMutation(t *datapb.CompactionTask, resul
...
-			DmlPosition: getMinPosition(lo.Map(latestCompactFromSegments, func(info *SegmentInfo, _ int) *msgpb.MsgPosition {
-				return info.GetDmlPosition()
-			})),
+			StartPosition:       getMinPosition(compactFromSegInfos),
+			DmlPosition:         getMinPosition(compactFromSegInfos),
 		}
 		segment := NewSegmentInfo(segmentInfo)
-		newSegments = append(newSegments, segment)
+		compactToSegInfos = append(compactToSegInfos, segment)
+		compactToSegIDs = append(compactFromSegIDs, segment.GetID())
```
Comment (czs007) [inline internal/datacoord/meta.go:1421]: compactToSegIDs = append(compactToSegIDs, segment.GetID())

File: internal/datacoord/meta.go [hunk 2]
```diff
@@ -1367,92 +1391,81 @@ func (m *meta) completeClusterCompactionMutation(t *datapb.CompactionTask, resul
...
 			LastExpireTime:      tsoutil.ComposeTSByTime(time.Unix(t.GetStartTime(), 0), 0),
 			Level:               datapb.SegmentLevel_L2,
-			StartPosition: getMinPosition(lo.Map(latestCompactFromSegments, func(info *SegmentInfo, _ int) *msgpb.MsgPosition {
-				return info.GetStartPosition()
-			})),
-			DmlPosition: getMinPosition(lo.Map(latestCompactFromSegments, func(info *SegmentInfo, _ int) *msgpb.MsgPosition {
-				return info.GetDmlPosition()
-			})),
+			StartPosition:       getMinPosition(compactFromSegInfos),
+			DmlPosition:         getMinPosition(compactFromSegInfos),
```
Comment (czs007) [inline internal/datacoord/meta.go:1417]: should use DmlPosition

File: internal/datacoord/meta.go (hunk 1 above)
Comment (wayblink) [reply internal/datacoord/meta.go]: resolved

File: internal/datacoord/meta.go (hunk 2 above)
Comment (wayblink) [reply internal/datacoord/meta.go]: resolved

**Pull request merged**